### PR TITLE
Request function of bridge contract 

### LINKED ISSUE #1331: Request function of bridge contract
- [ ] Define state and request function to store requests and emit event from other contracts.

### PR DESCRIPTION
**Pull request by Benzbeeb**
fixed #1331

### CODE REVIEW

File: bridges/evm/contracts/BridgeV2.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
+pragma solidity 0.5.17;
+pragma experimental ABIEncoderV2;
+
+import {Bridge} from "./Bridge.sol";
+import {IBridgeV2} from "./BridgeV2.sol";
+
+
+contract BridgeV2 is Bridge, IBridgeV2 {
+    struct Request {
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeV2.sol]: Let's reuse the same request packet struct

File: bridges/evm/contracts/BridgeV2.sol [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        returns (bytes32)
+    {
+        emit RequestOracle(
+            request.clientId,
+            request.oracleScriptId,
+            request.calldata,
+            request.askCount,
+            request.minCount
+        );
+        bytes32 key = keccak256(abi.encodePacked(msg.sender, request.clientId));
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeV2.sol:30]: Should fail on duplicated key.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/BridgeV2.sol (hunk 1 above)
Comment (taobun) [reply bridges/evm/contracts/BridgeV2.sol]: Noooooooo. It mean use RequestPacket struct in `IBridge.sol`

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/BridgeV2.sol [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+contract BridgeV2 is Bridge, IBridgeV2 {
+    struct PendingRequest {
+        IBridge.RequestPacket request;
+        address targetAddress;
+        bool isValue;
+    }
+
+    mapping(bytes32 => PendingRequest) public pendingRequest;
+    event RequestOracle(
```
Comment (taobun) [inline bridges/evm/contracts/BridgeV2.sol]: In IBridgeV2 already declare this event, we need to re-declare this?

File: bridges/evm/contracts/BridgeV2.sol [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        emit RequestOracle(
+            request.clientId,
+            request.oracleScriptId,
+            request.params,
+            request.askCount,
+            request.minCount
+        );
+        bytes32 key = keccak256(abi.encodePacked(msg.sender, request.clientId));
+        if (pendingRequest[key].isValue) revert();
```
Comment (taobun) [inline bridges/evm/contracts/BridgeV2.sol]: require(!pendingRequest[key].isValue, "DUPLICATE_PENDING_REQUEST");

File: bridges/evm/contracts/BridgeV2.sol [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        IBridge.RequestPacket memory r = IBridge.RequestPacket(
```
Comment (taobun) [inline bridges/evm/contracts/BridgeV2.sol]: I think we can just write IBridge.RequestPacket memory temp = request;

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/IBridge.sol [hunk 6]
```diff
@@ -7,7 +7,7 @@ interface IBridge {
     struct RequestPacket {
         string clientId;
         uint64 oracleScriptId;
-        string params;
+        bytes params;
```
Comment (taobun) [inline bridges/evm/contracts/IBridge.sol:10]: Use string for now

**Review by taobun — CHANGES_REQUESTED** [review]

**Pull request closed without merging**